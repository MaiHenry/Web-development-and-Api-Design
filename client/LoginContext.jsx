import React, { useState, useEffect } from "react";

// Implementation of context file for Login
export const LoginContext = React.createContext({
  user: undefined,
  profileName: undefined,
  applicationConfig: undefined,
  loginMethod: undefined,
  loadUser: () => {},
  unloadUser: () => {},
});

// Provider & states for information
export const LoginProvider = ({ children }) => {
  const [user, setUser] = useState(undefined);
  const [profileName, setProfileName] = useState(undefined);
  const [loginMethod, setLoginMethod] = useState(undefined);

  // Configuration for different auth.
  const applicationConfig = {
    google: {
      client_id: process.env.GOOGLE_CLIENT_ID,
      openid_configuration: process.env.GOOGLE_OPENID_CONFIGURATION,
    },
    microsoft: {
      client_id: process.env.MICROSOFT_CLIENT_ID,
      openid_configuration: process.env.MICROSOFT_OPENID_CONFIGURATION,
    },
  };

  const loadUser = async () => {
    try {
      const userId = window.sessionStorage.getItem("userId");
      const username = window.sessionStorage.getItem("username");
      
      if (userId && username) {
        setUser({ direct: { _id: userId, username } });
        setProfileName(username);
      } else {
        // If user data is not in SS, fetch from API
        const response = await fetch("/api/login");
        if (!response.ok) {
          throw new Error("Failed to fetch user " + response.statusText);
        }
        const userData = await response.json();

        if (userData.user.google) {
          setUser(userData.user.google); // Google users
          setProfileName(userData.user.google.name);
          setLoginMethod('google');
        } else if (userData.user.microsoft) {
          // Micrsoft users
          setUser(userData.user.microsoft);
          setProfileName(userData.user.microsoft.name);
          setLoginMethod('microsoft');
        } 
      }
    } catch (error) {
      console.error("Error loading user:", error);
    }
  };

  // Clear user data from sessionStorage and reset state
  const unloadUser = () => {
    setUser(undefined);
    setProfileName(undefined);
  };

  // useEffect to load user data
  useEffect(() => {
    loadUser();
  }, []);

  // Provider component
  return (
    <LoginContext.Provider
      value={{
        user,
        setUser,
        profileName,
        loginMethod,
        setProfileName,
        applicationConfig,
        loadUser,
        unloadUser,
      }}
    >
      {children}
    </LoginContext.Provider>
  );
};
