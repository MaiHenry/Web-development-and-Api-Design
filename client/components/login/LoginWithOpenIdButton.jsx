import React, { useEffect, useState } from "react";
import googleLogo from "../images/google_logo.png";
import microsoftLogo from "../images/microsoftlogin.png";
import { fetchJSON } from "../apiRequests/fetchJSON.jsx";
import "../../Styles.css";

export function LoginWithOpenIdButton() {
  const [googleAuthUrl, setGoogleAuthUrl] = useState("");
  const [microsoftAuthUrl, setMicrosoftAuthUrl] = useState("");

  // Load Google authorization URL with required parameters
  async function loadGoogleAuthorizationUrl() {
    const config = await fetchJSON("/auth/google/config");
    const params = {
      response_type: "token",
      scope: "email profile",
      client_id: config.client_id,
      redirect_uri: window.location.origin + "/login/callback",
      state: randomString(50),
    };

    setGoogleAuthUrl(
      `${config.authorization_endpoint}?${new URLSearchParams(params)}`
    );
  }

  // Load Microsoft authorization URL with required parameters
  async function loadMicrosoftAuthorizationUrl() {
    const config = await fetchJSON("/auth/microsoft/config");
    const redirect_uri = window.location.origin + "/login/callback";
    const code_verifier = randomString(50);
    window.sessionStorage.setItem("code_verifier", code_verifier);
    const code_challenge = await sha256(code_verifier);
    const state = randomString(50);

    const params = new URLSearchParams({
      response_mode: "fragment",
      response_type: "code",
      scope: "openid email profile",
      client_id: config.client_id,
      redirect_uri,
      code_challenge,
      code_challenge_method: "S256",
      state,
    });

    setMicrosoftAuthUrl(
      `${config.authorization_endpoint}?${params.toString()}`
    );
  }

  // Load authorization URLs when component mount
  useEffect(() => {
    loadGoogleAuthorizationUrl();
    loadMicrosoftAuthorizationUrl();
  }, []);

  return (
    <div>
      <div>
        <a href={googleAuthUrl}>
          <img
            className="googlelogo"
            src={googleLogo}
            alt="Google login"
            style={{ width: "200px" }}
          />
        </a>
      </div>
      <div>
        <a href={microsoftAuthUrl}>
          <img
            className="microsoftlogo"
            src={microsoftLogo}
            alt="Microsoft login"
            style={{ width: "200px" }}
          />
        </a>
      </div>
    </div>
  );
}

// Generate random string of length
function randomString(length) {
  let result = "";
  const characters =
    "ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789";
  const charactersLength = characters.length;
  for (let i = 0; i < length; i++) {
    result += characters.charAt(Math.floor(Math.random() * charactersLength));
  }
  return result;
}

// Calculate SHA-256 hash of a string
export async function sha256(string) {
  const binaryHash = await crypto.subtle.digest(
    "SHA-256",
    new TextEncoder("utf-8").encode(string)
  );
  return btoa(String.fromCharCode.apply(null, new Uint8Array(binaryHash)))
    .split("=")[0]
    .replace(/\+/g, "-")
    .replace(/\//g, "_");
}
