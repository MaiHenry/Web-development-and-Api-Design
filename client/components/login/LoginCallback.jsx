import React, { useContext, useEffect, useState } from "react";
import { useNavigate } from "react-router-dom";
import { LoginContext } from "../../LoginContext";
import { fetchJSON } from "../apiRequests/fetchJSON";

export function LoginCallback() {
  const { loadUser } = useContext(LoginContext);
  const { applicationConfig } = useContext(LoginContext);
  const [error, setError] = useState(null);
  const [errorDescription, setErrorDescription] = useState(null);
  const navigate = useNavigate();

  async function handleCallback() {
    const hash = Object.fromEntries(
      new URLSearchParams(window.location.hash.substring(1))
    );

    let { access_token, error, error_description, state, code } = hash;
    const isMicrosoftLogin = state && code; // Check if auth is microsoft.

    if (isMicrosoftLogin) {
      setError(error);
      setErrorDescription(error_description);

      if (code) {
        const microsoftConfig = applicationConfig.microsoft;
        const { token_endpoint } = await fetchJSON(
          microsoftConfig.openid_configuration
        );
        const code_verifier = window.sessionStorage.getItem("code_verifier");
        const redirect_uri = `${window.location.origin}/login/callback`;

        const res = await fetch(token_endpoint, {
          method: "POST",
          body: new URLSearchParams({
            grant_type: "authorization_code",
            code,
            client_id: microsoftConfig.client_id,
            code_verifier,
            redirect_uri,
          }),
        });
        const json = await res.json();
        access_token = json.access_token;
        console.log("Received access_token:", access_token); // Log the recieved token
      }
    }

    // Common logic for both Google and Microsoft
    if (access_token) {
      const loginUrl = `/api/login/${
        isMicrosoftLogin ? "microsoft" : "google"
      }`;

      const response = await fetch(loginUrl, {
        method: "POST",
        body: JSON.stringify({ access_token }),
        headers: {
          "Content-Type": "application/json",
        },
      });
      if (!response.ok) {
        throw new Error("Login failed " + response.statusText);
      }
      await loadUser();
      navigate("/");
    }
  }

  // Trigger login callback when component mounts
  useEffect(() => {
    handleCallback();
  }, []);

  if (error) {
    return (
      <>
        <h1>Error</h1>
        <div>{error}</div>
        {errorDescription && <p>{errorDescription}</p>}
      </>
    );
  } else {
    return <h2>Please wait ...</h2>;
  }
}
