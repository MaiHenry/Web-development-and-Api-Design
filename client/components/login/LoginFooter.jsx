import { Link, useNavigate } from "react-router-dom";
import React, { useContext } from "react";
import { LoginContext } from "../../LoginContext";
import "../../Styles.css";

export function LoginFooter() {
  const navigate = useNavigate();
  const { user, profileName, unloadUser } = useContext(LoginContext);

  async function handleSubmitLogout(e) {
    e.preventDefault();
    const res = await fetch("/api/login", {
      method: "DELETE",
    });
    if (!res.ok) {
      throw new Error("Failed to log out " + res.statusText);
    }
    unloadUser();
    navigate("/");
  }

  let profileStatus;

  if (profileName) {
    profileStatus = (
      <footer>
        Logged in as:{" "}
        <Link className="footerLink" to={"/profile"}>
          {profileName}
        </Link>
        <p>
          <button className="footerButton" onClick={handleSubmitLogout}>
            Log out
          </button>
        </p>
      </footer>
    );
  } else {
    profileStatus = <div>Log in to explore features!</div>;
  }

  return profileStatus;
}
