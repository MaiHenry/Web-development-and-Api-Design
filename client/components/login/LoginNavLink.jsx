import { Link } from "react-router-dom";
import React, { useContext } from "react";
import { LoginContext } from "../../LoginContext";

export function LoginNavLink() {
  const { user } = useContext(LoginContext);

  // Check if user is logged in
  if (user) {
    return <Link to="/profile">My Profile</Link>;
  }

  return <Link to="/login">Log in</Link>;
}
