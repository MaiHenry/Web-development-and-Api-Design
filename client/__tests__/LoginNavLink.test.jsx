import React from "react";
import renderer, { act } from "react-test-renderer";
import { MemoryRouter } from "react-router-dom";
import { LoginContext } from "../LoginContext";
import { LoginNavLink } from "../components/login/LoginNavLink";

describe("LoginNavLink", () => {
  it("renders when user logged in", () => {
    const user = { email: "ola@nordmannen.com" };

    let component;
    act(() => {
      component = renderer.create(
        <MemoryRouter>
          <LoginContext.Provider value={{ user }}>
            <LoginNavLink />
          </LoginContext.Provider>
        </MemoryRouter>,
      );
    });

    expect(component).toMatchSnapshot();
  });

  it("no render when not logged in", () => {
    let component;
    act(() => {
      component = renderer.create(
        <MemoryRouter>
          <LoginContext.Provider value={{ user: null }}>
            <LoginNavLink />
          </LoginContext.Provider>
        </MemoryRouter>,
      );
    });

    expect(component.toJSON()).toBeNull();
  });
});
