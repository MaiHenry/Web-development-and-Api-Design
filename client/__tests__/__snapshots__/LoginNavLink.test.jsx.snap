// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LoginNavLink renders when user logged in 1`] = `
<nav>
  <a
    className="navbar-link"
    href="/"
    onClick={[Function]}
  >
    Browse rooms
  </a>
  <a
    className="navbar-link"
    href="/people"
    onClick={[Function]}
  >
    People
  </a>
  <a
    className="navbar-link"
    href="/About"
    onClick={[Function]}
  >
    About
  </a>
  <a
    className="navbar-link"
    href="/profile"
    onClick={[Function]}
  >
    My Profile
  </a>
</nav>
`;
